proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,726 @@
+
+* 1C/.../5A: Not applicable.
+
+* 1C/.../5B: Not applicable.
+
+* 1C/.../5D: Not applicable.
+
+* 1C/2A/3C/4A/5C: **Allowed**: ``foo :: _ -> Int; foo True = 5``, with ``-XPartialTypeSignatures -Wno-partial-type-signatures``. Retained
+  with this proposal.
+
+* 1C/2A/3C/4B/5C: **Allowed**: ``foo :: _t -> _t; foo _ = True``, with ``-XPartialTypeSignatures -XNamedWildCards -Wno-partial-type-signatures``.
+  Named missing bits always print a diagnostic with this proposal.
+
+* 1C/2BC/3A/4A/5C: **Allowed**: Like 1C/2A/3C/4A/5C, but with different flags. Use a name (4B) to get a diagnostic with this
+  proposal.
+
+* 1C/2BC/3A/4B/5C: **Allowed**: Like 1C/2A/3C/4B/5C, but with different flags. This proposal removes the
+  dependency on flags.
+  
+* 1C/2BC/3B/4A/5C: **Not allowed**: ``foo :: Proxy _ -> (); foo (Proxy :: Proxy @Bool a) = ()``, but I want the know that type of the ``_`` is ``Bool``.
+  Use a name (4B) to get a diagnostic with this propopsal.
+
+* 1C/2BC/3B/4B/5C: **Not allowed**: Similar to 1C/2BC/3B/4A/5C. Allowed by this proposal.
+
+There are several problems not really reflected by this taxonomy:
+
+a. We might usefully want to combine several possibilities in close proximity. Changing between modes with module-wide flags is unpleasant.
+
+b. As discussed in motivation point (5), we won't want the choice between complete and partial type signatures to be controlled by
+   the presence of underscores.
+  
+Proposed Change Specification
+-----------------------------
+
+1. Outside of patterns, treat ``_`` and ``__`` as elisions everywhere. This means that ``_`` and ``__`` mean "I don't care".
+
+   - In types (with ``-XElidedTypes``), a ``_`` is treated as a fresh unification variable. This means that ``foo :: _ -> _`` is the same as
+     ``foo :: a -> b``, while ``Proxy @_ True`` is the same as ``Proxy @Bool True``. You're instructing GHC that
+     you want it to fill in the ``_`` with what is necessary for the type to kind-check. Once GHC is finished processing
+     the type, however, any opportunity to solve for ``_`` has been taken; at that point, if it is still unconstrained,
+     GHC generalizes over it, like it would a fresh normal type variable.
+
+     Elisions can appear anywhere a type can be written, but they cannot
+     appear as a constraint in a complete type signature. They do not cause
+     diagnostics to be printed. The existence of an elision does *not* cause a
+     signature to be treated as partial.
+
+     Without ``-XElidedTypes``, an underscore in a type is an error; the error message would suggest
+     either using a named wildcard to get a diagnostic or enabling ``-XElidedTypes`` to accept the
+     elision.
+
+     Each elision spelled ``__`` will cause GHC to print out a diagnostic warning explaining how
+     the elision was filled. Writing an elision as ``_`` suppresses the diagnostic, as does
+     ``-Wno-elided-types``.
```

Comment:
This feels overly complex. I suggest instead:

1. Type elisions are added to the language.
2. A new warning `-Welided-types` is added that emits a warning with the inferred type. Disabled by default, but part of `-Wall`.
3. The warning can be disabled with `-Wno-elided-types`.